### DISCUSSION TITLE
Get type of item in optional array

### DISCUSSION THREAD

**Discussion by winsvold** [ricokahler/sanity-codegen#196] [Q&A] [answered]
Hi, love this project 😄 

I've used the code gen to generate types, and one of my generated document types looks like this:
```
export interface MyDocument extends SanityDocument {
  _type: "events";
  events?: Array<
    SanityKeyed<{
      _type: "event";
      time?: string;
      description?: string;
    }>
  >;
}
```

Now in my app I want to create a component that renders one single event, and I want to extract the type like this:
```
function MyComponent(props: MyDocument['events'][number]) {
  return (
    <div>
      {props.time}
      {props.description}
    </div>
  )
}
```

But since the MyDocument['events']-array is optional it doesn't work. Do you have any suggestions as to how I can dig out this type? 

The error I get looks something like this: 
```
Type '({ _type: "event"; time?: string | undefined; description?: string | undefined; } & { _key: string; })[] | undefined' has no matching index signature for type 'number'.ts(2537)
```

**Comment by ricokahler** [comment, marked as answer]
hi there!, thanks for the kind words.

I have two potential solutions for you:
1. you could go into your schema field and add [`codegen: { required: true }`](https://github.com/ricokahler/sanity-codegen/tree/f69ba953552f832a0e1e6b64d642b3223599e346#schema-codegen-options). This will make the codegen assert your field as required (this will not be enforced so be careful to [add the corresponding validation too](https://www.sanity.io/docs/validation#4dc8b38bc411).

```ts
export default {
  name: 'myDocument',
  type: 'document',
  fields: [
    {
      name: 'aRequiredField',
      type: 'string',
      // 👇👇👇
      codegen: { required: true },
      validation: (Rule) => Rule.required(),
      // 👆👆👆
    },
  ],
};
```

2. you could create a typescript helper that allows you to reach into the object. I have made helpers like this before.

```ts
type MyDocument = { events?: Array<{_type: 'something'}> }

type SafeIndexedAccess<
    T extends { [key: string]: any } | undefined,
    K extends keyof NonNullable<T>,
  > = T extends undefined ? NonNullable<T>[K] | undefined : NonNullable<T>[K];

type Props = SafeIndexedAccess<MyDocument['events'], number>
// results in: type Props = { _type: 'something'; } | undefined
```

[ts playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAsiAiB7AxgVwLYQHbCgXigG8oIA3bYAZwH4AuKAQQCcmBDEAHkIH1RJ6A5JUSZgACwCWWAOYCAvgD4ocgFAq+0AMqsAZhACSWACYQAHhCMNkyCJUocVUJ1AAqJU8GxHKRKAG0AawgQekpgJilpAF16ViwQZSgAHyhUYwgdKQsAGkdnAGl3T2MfIJBEHSgAOUQsKtQAGwbWACMGiA4XBVynJQI3M2LvVPTMrAsoamra+qbW9s6FP3yo5JGTMYn6GrrG5raOruWogG41DSgABSZEMB8CbT1DE3NLa1t7OCQ0TBw-ATIFEoAii2SgWAwLQgTAUKgA9HCoExbI0qFApPQLtdbvdfLxwBBBMJRJIZAITokUmkNlkjCogA)

or without the helper 🤷:

```ts
type Props = NonNullable<MyDocument['events']> | undefined;
```

**Reply by winsvold** [reply]
Wow that was fast, the SafeIndexedAccess-helper was exactly what I needed, thank you so much! 

And impressive typescript skills, kudos 🥳
